### ISSUE TITLE
Application crashes when state loss is caused by closing a Workbook

### ISSUE THREAD

**Issue by guwidoe** [cristianbuse/VBA-StateLossCallback#3]
I'm writing this issue for the second time because for some reason the first time it didn't go through when I submitted it.

The application crashes when I set up a state loss call back and then close the workbook containing it by using the top right "x" button.

```vba
Sub SetUpStateLossCallBack()
    Static stateTracker As StateLossCallback
    Set stateTracker = New StateLossCallback
    stateTracker.InitByMacroName "TestStateLossCall"
    Debug.Print "Initialized state loss callback"
End Sub

Sub TestStateLossCall()
    Debug.Print "Called back"
    
    Dim testFile As String
    testFile = CreateObject("WScript.Shell").SpecialFolders("Desktop") & "\TestFile.txt"
    Dim fileNum As Long: fileNum = FreeFile()
    
    Open testFile For Output Access Write As #fileNum
        Print #fileNum, "Called back"
    Close #fileNum
End Sub
```

This is the code I used to verify the callback was never successfully executed.

It works fine if the state loss is caused by the stop button in the VBE.

The crash also happens with my implementation of the release trick implemented in the [`KeepRecalculating`](https://gist.github.com/guwidoe/f248271da14d40a6ec7b44ccca245479) function.

I remember this working in the past, so I wonder if you can reproduce the crash or have any idea why it happens.

**Comment by cristianbuse** [comment]
Yes, it crashes for me also. It definitely worked when I first posted this. Even the ```IUnknown::Release``` was working when closing the book. I have no idea why - something must have changed since then. 

Interestingly, when I put a breakpoint on ```DataObject_Clear``` the code actually stops in debug mode and only then the crash occurs. Not sure if this was happening before.

Anyway, ```InitByAddress``` still works without crashing. So, it seems ```Application.Run``` is the culprit here.

**Comment by thefasquelle** [comment]
My workaround for this issue is to cancel the callback in the Workbook_BeforeClose event as explained in #4 

However 

> Anyway, `InitByAddress` still works without crashing. So, it seems `Application.Run` is the culprit here.

On my computer, InitByAddress is very inconsistent. Sometimes it works and sometimes it crashes (not only on workbook close). Whereas InitByMacroName crashes only when workbook is closed, which can be worked around.

**Comment by cristianbuse** [comment]
Thanks @thefasquelle for the feedback!

I would obviously want to find a workaround that does not have a dependency on a Workbook event. However, I've been extremely busy in the last few months and had no time to work on this. Hopefully I can spend some time in the next few weeks.
